### DISCUSSION TITLE
return value of IObservableCache SortAndBind

### DISCUSSION THREAD

**Discussion by aguahombre** [reactivemarbles/DynamicData#958] [Q&A]
Should the ```IObservable<IChangeSet<TObject, TKey>>``` returned by the ```IObservableCache SortAndBind``` be in sort order?
I assumed it would be, but calling ```QueryWhenChanged()``` on the returned observable produces an unordered list.

**Comment by JakenVeina** [comment]
My intuition says "no, they should not be ordered".

The functional goal of the operator, to me, is only to subscribe to and consume changesets, to apply them to a target, not to actually perform transformations. Me personally, I would have clarified this fact by having the operator return `IDisposable` or `IObservable<Unit>`.

Looking at the code, what the operator is doing is effectively...

```cs
return source.Do(...);
```

So, yeah, it's quite literally just passing through the changes, for convenience.

However.

I ALSO note that the code, internally is ACTUALLY doing the following...

```cs
_sorted = source.Do(...);

...

return _sorted;
```

So, it's entirely possible that @RolandPheasant intended to have the downstream be sorted. I suspect that's not the case, because the big point of combining `.Sort()` and `.Bind()` together was for optimization, and pretty much all of that optimization would be lost if the operator had to output sorted changesets. Internally, it would have to treat sorting as a separate operation from binding, in order to capture changesets representing the sort operation, which is precisely what .`Sort()` does.

If @RolandPheasant can confirm, we can definitely update documentation to clarify that the downstream is a passthrough for convenience only.

**Reply by aguahombre** [reply]
Thank you for your detailed answer.
It makes sense that if you could generate a ```IChangeSet``` in sort order cheaply, you wouldn't need the sort and bind method.

An update to the documentation would be useful.

I notice that the top level documentation for ```ObservableCache``` still suggests using the obsolete ```.Sort()``` method.

**Comment by RolandPheasant** [comment]
> So, it's entirely possible that @RolandPheasant intended to have the downstream be sorted

Regarding the cache side of DD, sorting was only ever intended for binding and virtualization.  The reason is due to the limitation of storing the sate in a dictionary which has no index  awareness.  I recall originally, I tried to store sorted state in a sorted dictionary, which failed. Then I tried to use an internal container which held an index. This also failed, and I realized it was pointless trying to keep the state sorted across downstream operators would exponentially complicate an already difficult problem domain.
